### PR TITLE
Simulation Side Rivers

### PR DESCRIPTION
**Pull request by BrickPi**
Loads the rivers bitmap into structure that the sim can eventually use for movement and combat checks. No graphics in this PR, probably going to shelve working on river graphics and focus on sim-side as I get more availability to contribute again.

Tested, loads 905 rivers successfully in vanilla.

### CODE REVIEW

File: src/openvic-simulation/map/MapDefinition.hpp [hunk 1]
```diff
@@ -20,6 +21,16 @@ namespace OpenVic {
...
 	struct ModifierManager;
 
+	struct RiverSegment {
+		friend struct MapDefinition;
+
+	private:
+			const uint8_t PROPERTY(size);
+			const std::vector<ivec2_t> PROPERTY(points);
+
+			RiverSegment(uint8_t new_size, std::vector<ivec2_t> new_points);
```
Comment (Hop311) [inline src/openvic-simulation/map/MapDefinition.hpp:31]:
```suggestion
		const uint8_t PROPERTY(size);
		std::vector<ivec2_t> PROPERTY(points);

		RiverSegment(uint8_t new_size, std::vector<ivec2_t>&& new_points);

	public:
		RiverSegment(RiverSegment&&) = default;
```
- Make `points` non-const so it can be moved (const vectors can only be copied).
- For `new_points` to be a r-value reference/moved.
- Add a public default move constructor
- Indent back a tab to be one past private.

File: src/openvic-simulation/map/MapDefinition.cpp [hunk 2]
```diff
@@ -12,6 +14,8 @@ using namespace OpenVic::NodeTools;
 
 MapDefinition::MapDefinition() : dims { 0, 0 }, max_provinces { ProvinceDefinition::MAX_INDEX } {}
 
+RiverSegment::RiverSegment(uint8_t new_size, std::vector<ivec2_t> new_points) : size { new_size }, points { std::move(new_points) } {}
```
Comment (Hop311) [inline src/openvic-simulation/map/MapDefinition.cpp:17]:
```suggestion
RiverSegment::RiverSegment(uint8_t new_size, std::vector<ivec2_t>&& new_points)
	: size { new_size }, points { std::move(new_points) } {}
```
Make `new_points` an r-value reference and move the initialiser list to the next line

File: src/openvic-simulation/map/MapDefinition.cpp [hunk 3]
```diff
@@ -519,18 +525,33 @@ bool MapDefinition::load_map_images(fs::path const& province_path, fs::path cons
...
-	if (terrain_bmp.get_bits_per_pixel() != expected_terrain_bpp) {
+	if (terrain_bmp.get_bits_per_pixel() != expected_terrain_rivers_bpp) {
+		Logger::error(
+			"Invalid terrain BMP bits per pixel: ", terrain_bmp.get_bits_per_pixel(), " (expected ", expected_terrain_rivers_bpp, ")"
+		);
+		return false;
+	}
+
+	BMP rivers_bmp;
+	if (!(rivers_bmp.open(rivers_path) && rivers_bmp.read_header() && rivers_bmp.read_palette() && rivers_bmp.read_pixel_data())) {
```
Comment (Hop311) [inline src/openvic-simulation/map/MapDefinition.cpp:536]: Is `read_palette` necessary here? We're only using the indices in the pixel data from what I can see

File: src/openvic-simulation/map/MapDefinition.cpp [hunk 4]
```diff
@@ -637,6 +658,183 @@ bool MapDefinition::load_map_images(fs::path const& province_path, fs::path cons
...
+		}
+		// check pixel to left
+		if (last_segment_end.x > 0 && last_segment_direction != 3) {
+			if (river_data[idx - 1] < 12) {
+				points.push_back({ last_segment_end.x - 1, last_segment_end.y });
+				direction = 4;
+			}
+		}
+
+		uint8_t size = river_data[points.front().x + points.front().y * rivers_bmp.get_width()] - 1; // size of river from 1 - 10 determined by colour
```
Comment (Hop311) [inline src/openvic-simulation/map/MapDefinition.cpp:727]: This line could crash if `points` is empty (e.g. a standalone starting point). We need to detect that error here and fail gracefully. Also the `uint8_t direction` variable could contain random data unless it's explicitly initialised, safest to set it to 0 even if we're sure it'll get initialised by later code.

File: src/openvic-simulation/map/MapDefinition.cpp [hunk 5]
```diff
@@ -637,6 +658,183 @@ bool MapDefinition::load_map_images(fs::path const& province_path, fs::path cons
...
+			if (limit == 4096) {
+				Logger::error("River segment starting at (", points.front().x, ", ", points.front().y, ") is longer than limit 4096, check for misplaced pixels or other definition errors!");
+				river_complete = true;
+				break;
+			}
+
+			idx = points.back().x + points.back().y * rivers_bmp.get_width();
+
+			ivec2_t merge_location;
+			bool merge;
```
Comment (Hop311) [inline src/openvic-simulation/map/MapDefinition.cpp:745]: This variable could be randomly true or false if not explicitly initialised (with MSVS at least), so it needs to be `bool merge = false;`

File: src/openvic-simulation/map/MapDefinition.cpp [hunk 6]
```diff
@@ -637,6 +658,183 @@ bool MapDefinition::load_map_images(fs::path const& province_path, fs::path cons
...
+			// no further points
+			if (merge) points.push_back(merge_location);
+			river_complete = true;
+			break;
+		}
+
+		// TODO: implement optimisation algorithm to remove irrelevant points. Ramer-Douglas-Peucker?
+
+		// add segment then recursively call if neeeded
+		river.push_back({ size, points });
```
Comment (Hop311) [inline src/openvic-simulation/map/MapDefinition.cpp:817]:
```suggestion
		river.push_back({ size, std::move(points) });
```
Move the `points` vector.

File: src/openvic-simulation/map/MapDefinition.cpp [hunk 7]
```diff
@@ -637,6 +658,183 @@ bool MapDefinition::load_map_images(fs::path const& province_path, fs::path cons
...
+
+		if (river_complete) return;
+		next_segment(new_point, direction, river);
+	};
+
+	// find every river source and then run the segment algorithm.
+	int temp = 0;
```
Comment (Hop311) [inline src/openvic-simulation/map/MapDefinition.cpp:824]: Leftover `temp` variable to remove?

File: src/openvic-simulation/map/MapDefinition.cpp [hunk 8]
```diff
@@ -637,6 +658,183 @@ bool MapDefinition::load_map_images(fs::path const& province_path, fs::path cons
...
+	for (int y = 0; y < rivers_bmp.get_height(); ++y) {
+		for (int x = 0; x < rivers_bmp.get_width(); ++x) {
+			if (river_data[x + y * rivers_bmp.get_width()] == START_COLOUR) { // start of a river
+				river_t river;
+
+				next_segment({ x, y }, 0, river);
+
+				rivers.push_back(river);
```
Comment (Hop311) [inline src/openvic-simulation/map/MapDefinition.cpp:832]:
```suggestion
				rivers.push_back(std::move(river));
```
Move the `river`

**Review by Hop311 — COMMENTED** [review]
Great work, some minor code comments but the algorithm looks good. Could you test edge cases/invalid river formations just in case there are any other potential crash bugs? In the case of invalid data it doesn't matter much about loading the rest, just gracefully exiting the function without segfaulting

**Comment by Hop311** [comment]
Also rather than storing every single point of each river, we could just store the corner points (so replace the last point if `direction == last_segment_direction`). This would require us to store the start point of the river as well (which I don't think we store at all atm?), but would save us a lot of memory/later processing time.

Also I'm wondering if it's necessary to have separate code for finding the first pixel vs the later pixels of a river segment? Could we find a way to handle all pixel with the same "next pixel finding" code?

File: .gitignore [hunk 9]
```diff
@@ -36,7 +36,7 @@
 
 # VSCode
 .vscode/*
-!.vscode/launch.json
+.vscode/launch.json
```
Comment (Hop311) [inline .gitignore:39]: Did you intend to change this? The `!` means `.vscode/launch.json` isn't ignored despite the `.vscode/*` line above. If we should be ignoring the file then the line should be removed altogether as the line above will already catch it.

File: src/openvic-simulation/map/MapDefinition.cpp [hunk 10]
```diff
@@ -637,6 +659,196 @@ bool MapDefinition::load_map_images(fs::path const& province_path, fs::path cons
...
+
+			// no further points
+			if (merge) points.push_back(merge_location);
+			river_complete = true;
+			break;
+		}
+
+		// save memory & simplify by storing only start, corner, and end points.
+		const auto is_corner_point = [](ivec2_t previous, ivec2_t current, ivec2_t next) {
+			return ((current.x - previous.x) * (next.y - current.y)) != ((current.y - previous.y) * (next.x - current.x)); // slope is fun
```
Comment (Hop311) [inline src/openvic-simulation/map/MapDefinition.cpp:821]: Nice!

File: src/openvic-simulation/map/MapDefinition.cpp [hunk 11]
```diff
@@ -637,6 +659,196 @@ bool MapDefinition::load_map_images(fs::path const& province_path, fs::path cons
...
+		next_segment(new_point, direction, river);
+	};
+
+	// find every river source and then run the segment algorithm.
+	for (int y = 0; y < rivers_bmp.get_height(); ++y) {
+		for (int x = 0; x < rivers_bmp.get_width(); ++x) {
+			if (river_data[x + y * rivers_bmp.get_width()] == START_COLOUR) { // start of a river
+				river_t river;
+
+				next_segment({ x, y }, 0, river);
```
Comment (Hop311) [inline src/openvic-simulation/map/MapDefinition.cpp:844]: If I understand correctly the start point `{ x, y }` itself isn't included in the river, is it not needed for drawing the river or calculating crossings? If it is needed please include it in the river object

**Review by Hop311 — COMMENTED** [review]
Very elegent corner detection 😃

File: .gitignore (hunk 9 above)
Comment (BrickPi) [reply .gitignore:39]: Intended to change to prevent placing my local launch.json under source control, I'll remove the line.

File: src/openvic-simulation/map/MapDefinition.cpp (hunk 11 above)
Comment (BrickPi) [reply src/openvic-simulation/map/MapDefinition.cpp:844]: V2 doesn't appear to include the start pixel, 
image
image

**Review by Hop311 — APPROVED** [review]
